dlt/destinations/sql_client.py

```diff
@@ -299,18 +321,48 @@ def df(self, chunk_size: int = None, **kwargs: Any) -> Optional[DataFrame]:
         May use native pandas/arrow reader if available. Depending on
         the native implementation chunk size may vary.
         """
+        try:
+            return next(self.iter_df(chunk_size=chunk_size))
+        except StopIteration:
+            return None
+
+    def arrow(self, chunk_size: int = None, **kwargs: Any) -> Optional[ArrowTable]:
+        """Fetches results as data frame in full or in specified chunks.
+
+        May use native pandas/arrow reader if available. Depending on
+        the native implementation chunk size may vary.
+        """
+        try:
+            return next(self.iter_arrow(chunk_size=chunk_size))
+        except StopIteration:
+            return None
+
+    def iter_fetchmany(self, chunk_size: int) -> Generator[List[Tuple[Any, ...]], Any, Any]:
+        while True:
+            if not (result := self.fetchmany(chunk_size)):
+                return
+            yield result
+
+    def iter_df(self, chunk_size: int) -> Generator[DataFrame, None, None]:
         from dlt.common.libs.pandas_sql import _wrap_result
 
         columns = self._get_columns()
-        if chunk_size is None:
-            return _wrap_result(self.native_cursor.fetchall(), columns, **kwargs)
-        else:
-            df = _wrap_result(self.native_cursor.fetchmany(chunk_size), columns, **kwargs)
-            # if no rows return None
-            if df.shape[0] == 0:
-                return None
-            else:
-                return df
+
+        # if no chunk size, fetch all
+        if not chunk_size:
+            yield _wrap_result(self.fetchall(), columns)
+            return
+
+        # otherwise iterate over results in batch size chunks
+        for result in self.iter_fetchmany(chunk_size=chunk_size):
+            # TODO: ensure that this is arrow backed
+            yield _wrap_result(result, columns, dtype_backend="pyarrow")
+
+    def iter_arrow(self, chunk_size: int) -> Generator[ArrowTable, None, None]:
+        """Default implementation converts df to arrow"""
+        for df in self.iter_df(chunk_size=chunk_size):
+            # TODO: is this efficient?
```

Comment:
We're piping everything through Pandas here. Pandas is known for being slow. Perhaps these are faster:

- https://docs.pola.rs/api/python/stable/reference/api/polars.read_database.html
- https://docs.pola.rs/api/python/stable/reference/api/polars.read_database_uri.html#polars.read_database_uri

They return a Polars dataframe from a database connection. Polars is Arrow-backed. The Polars dataframe can be turned into an Arrow table with https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.to_arrow.html.

Reply:
I agree. We already code doing the above efficiently in `sql_database` and we'll use it here. There's once "complication" though:
1. if the relation is a table, we have a correct schema and we can create properly typed arrow table. that also applies to pandas above
2. if not we must infer types from the data
`sql_database` code handles the above. Again that should be a next ticket. Somehow we should be able to attach Relation schema

Reply:
This is implemented now, we have the same code that the sql database uses to create arrow tables from cursor results, in the case of querying full tables we can use the known dlt schema for proper typing, for all results from calling "query" the types are autodiscovered. We could take this further by:
* Somehow using sql glot to figure out which column came from which table
* Allowing filtering, sorting and selecting a subste of columns of a given table where we can then easily use the known schema. Maybe this could be a next step.